src/bin/pg_autoctl/cli_service.c

```diff
@@ -238,6 +211,50 @@ cli_monitor_run(int argc, char **argv)
 }
 
 
+/*
+ * ensure_monitor_running checks if monitor is running, attempts to restart it if it is not.
+ * verifies the extension version is the same as expected and returns the postgres uri for the
+ * monitor connection upon success.
+ */
+static void
+ensure_monitor_running(Monitor *monitor, MonitorConfig *mconfig)
+{
+	MonitorExtensionVersion version = { 0 };
+	char postgresUri[MAXCONNINFO];
+
+	if (!pg_is_running(mconfig->pgSetup.pg_ctl, mconfig->pgSetup.pgdata))
+	{
+		log_info("Postgres is not running, starting postgres");
+
+		if (!pg_ctl_start(mconfig->pgSetup.pg_ctl,
+						  mconfig->pgSetup.pgdata,
+						  mconfig->pgSetup.pgport,
+						  mconfig->pgSetup.listen_addresses))
+		{
+			log_error("Failed to start PostgreSQL, see above for details");
+			exit(EXIT_CODE_PGCTL);
```

Comment:
I think a lower-level function should (almost) never exit itself, but return false so that the caller can decide the best way forward, which might still be an exit.